### PR TITLE
Fix navigation product's collection

### LINKED ISSUE #618: ProductBundle:CollectionController calls non existent function 'getProductsByCollectionIdPager'
# Environment

#### Sonata packages

```
sonata-project/ecommerce  dev-master
```

## Subject
As reported in #322 the CollectionController calls the non-existend function getProductsByCollectionIdPager on the `Sonata\ProductBundle\Entity\ProductSetManager`.

In addition, the ProductSetManager class does not add any actual functionalities, it only throws some RuntimeExceptions for some of the inherited methods and all other calls to this manager are handled by the ProductManager.   So before creating a PR to fix this bug, it would help to know the purpose of the `Sonata\ProductBundle\Entity\ProductSetManager`. 

If no one missed this controller action for the past four/five years it may be dropped altogether?
If not this action could (almost) be handled with the use of the existing method `queryInSameCollections` I think, but that one is also in the ProductManager so no need for the ProductSetManager there. 

## Steps to reproduce
Call SonataProductBundle:Collection:listProducts

## Expected results
List of products within a collection

## Actual results
```
Fatal error: Call to undefined method Sonata\ProductBundle\Entity\ProductSetManager::getProductsByCollectionIdPager() in \vendor\sonata-project\ecommerce\src\ProductBundle\Controller\CollectionController.php 
```

### PR DESCRIPTION
**Pull request by salamichel**
## Subject

- display list of products' collection on product page
- each collection target to a dedicated page 
- on the collection page display all related products

Closes #618 #322

## Changelog

```markdown
### Fixed
- fix collection navigation
- remove deprecated calls
```

### CODE REVIEW

File: src/ProductBundle/Controller/CollectionController.php [hunk 1]
```diff
@@ -21,92 +21,57 @@
...
      *
      * @param $collectionId
-     * @param $slug
-     *
-     * @throws NotFoundHttpException
      *
      * @return Response
      */
-    public function viewAction($collectionId, $slug)
+    public function listProductsAction(Request $request, $collectionId = null)
```
Comment (wbloszyk) [inline src/ProductBundle/Controller/CollectionController.php]:
```suggestion
    public function index(Request $request)
```
It is simmilar to CatalogController:index, so make the same logic

File: src/ProductBundle/Controller/CollectionController.php [hunk 2]
```diff
@@ -21,92 +21,57 @@
...
+        $page = $request->get('page', 1);
+        $displayMax = $request->get('max', 9);
+        $displayMode = $request->get('mode', 'grid');
+        $option = $request->get('option');
+
+        if (!\in_array($displayMode, ['grid'], true)) { // "list" mode will be added later
+            throw new NotFoundHttpException(sprintf('Given display_mode "%s" doesn\'t exist.', $displayMode));
+        }
+
+        $collectionId = $request->get('collection_id');
```
Comment (wbloszyk) [inline src/ProductBundle/Controller/CollectionController.php]:
```suggestion
```

File: src/ProductBundle/Resources/config/routing/catalog.xml [hunk 3]
```diff
@@ -4,8 +4,14 @@
         <default key="_controller">SonataProductBundle:Catalog:index</default>
     </route>
     <route id="sonata_catalog_category" path="/{category_slug}/{category_id}">
+        <requirement key="category_id">\d+</requirement>
         <default key="_controller">SonataProductBundle:Catalog:index</default>
         <default key="category_id">0</default>
         <default key="category_slug">all</default>
     </route>
+    <route id="sonata_catalog_collection_products" path="/collection/{collection_id}-{collection_slug}">
+        <default key="_controller">SonataProductBundle:Collection:listProducts</default>
```
Comment (wbloszyk) [inline src/ProductBundle/Resources/config/routing/catalog.xml]:
```suggestion
        <default key="_controller">SonataProductBundle:Collection:index</default>
```

File: src/ProductBundle/Entity/ProductManager.php [hunk 4]
```diff
@@ -162,6 +162,21 @@ public function getPager(array $criteria, $page, $limit = 10, array $sort = [])
         return $pager;
     }
 
+    public function queryInSameCollection($collection, $limit = null)
```
[the remark this answers is not in the document]
Comment (jeroendk) [reply src/ProductBundle/Entity/ProductManager.php]: Also queryInSameCollection is already defined

File: src/ProductBundle/Resources/views/Product/view.html.twig [hunk 5]
```diff
@@ -82,6 +82,20 @@ file that was distributed with this source code.
                                     {% endblock %}
                                 </li>
                             {% endblock %}
+                            {% block product_collections %}
```
Comment (jeroendk) [inline src/ProductBundle/Resources/views/Product/view.html.twig]: Is this BC? Don't think you want new items in your existing layout when upgrading a minor version of this bundle? Also maybe move it to a dedicated Twig file?

File: src/ProductBundle/Entity/ProductManager.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (wbloszyk) [reply src/ProductBundle/Entity/ProductManager.php]: @jeroendk already define is queryInSameCollections. Method from this PR do something else.

[the remark this answers is not in the document]
Comment (jeroendk) [reply src/ProductBundle/Entity/ProductManager.php]: Yes I know.

[the remark this answers is not in the document]
Comment (salamichel) [reply src/ProductBundle/Entity/ProductManager.php]: queryInSameCollection**s** is already defined , i proposed queryInSameCollection to query products in one collection. i will rename to queryInCollection

**Comment by salamichel** [comment]
@greg0ire @core23  i don't get what's wrong with Travis ?! could you please help on this ?

**Comment by quisse** [comment]
@salamichel What's the reason behind the switch from `->` to `::` in your tests?

**Comment by salamichel** [comment]
Because travis failed so i tried something

**Comment by salamichel** [comment]
@quisse i reverted modification but still issue on 
1) Sonata\Component\Tests\Basket\BasketManagerTest::testGetPager
Error: Call to protected method PHPUnit\Framework\TestCase::createMock() from context 'Sonata\Doctrine\Test\EntityManagerMockFactory'

(...)

all PR failed due to this error

**Comment by quisse** [comment]
AFAIK there's something wrong with https://github.com/sonata-project/sonata-doctrine-extensions. `vendor/sonata-project/doctrine-extensions/src/Test/EntityManagerMockFactory.php` should extend `\PHPUnit\Framework\TestCase` instead of getting an instance via argument in order to make `createMock` accessible.

**Comment by salamichel** [comment]
I don't get it, all these tests passed last week. they havent not been modified. There are also already extends ` \PHPUnit\Framework\TestCase`

**Comment by greg0ire** [comment]
You will have to require the latest version of the sonata-doctrine-extensions to fix the prefer-lowest build

**Comment by salamichel** [comment]
@greg0ire Hi i tried to remove many deprecated call, but ii'm facing with issue on lowest-prefer and symfony 3.4*

Do my actions are it to soon ?

What is recommandation about using 
Symfony\Component\Intl\Currencies
Symfony\Contracts\EventDispatcher\Event

The method "Symfony\Component\Intl\Intl::getCurrencyBundle()" is deprecated since Symfony 4.3, use "Symfony\Component\Intl\Currencies" instead.
The "Sonata\Component\Event\AddBasketElementEvent" class extends "Symfony\Component\EventDispatcher\Event" that is deprecated since Symfony 4.3, use "Symfony\Contracts\EventDispatcher\Event" instead.

**Comment by greg0ire** [comment]
Maybe we can drop 3.4 if that helps, we are only supposed to support 1 LTS. Otherwise, we should find a way to address the deprecation conditionally, depending on the version of sf.

**Comment by wbloszyk** [comment]
@greg0ire 
Drop 3.4 (and support 4.4 || 5.0) is good idea for master branch (4.x version).

@salamichel
I think deprecated call should be removed in master branch to keep BC.

I resolve travis (test!) issue. See #632

**Comment by salamichel** [comment]
@greg0ire i'm almost ready for review, 

1. i have an issue for prefer-lowest on `Makefile:66: recipe for target 'test' failed` i don't know if it come from my modification?

2. Also i considering that 'SYMFONY=3.4.*' will be drop for 'SYMFONY=4.4.*' tell me if true or if i need to pass 'SYMFONY=3.4.*' tests ?

**Comment by greg0ire** [comment]
1. That happens when phpunit exits with a non-zero exit code. You can try to reproduce this locally by running `composer update --prefer-lowest && make test`
2. A PR needs to be done on https://github.com/sonata-project/dev-kit for your changes in travis.yaml to stay

**Comment by greg0ire** [comment]
1. A solution might be to require the phpunit bridge v5

**Comment by greg0ire** [comment]
1,. was adressed in #632

File: src/BasketBundle/Block/BasketBlockService.php [hunk 6]
```diff
@@ -24,7 +24,7 @@
 /**
  * @author Xavier Coureau <xcoureau@ekino.com>
  */
-class BasketBlockService extends AbstractAdminBlockService
+class BasketBlockService extends AbstractBlockService
```
Comment (wbloszyk) [inline src/BasketBundle/Block/BasketBlockService.php]: BasketBlockService does not imprement AdminBlockServiceInterface so it is BC-break. Anyway you should extends EditableBlockService.

File: src/BasketBundle/Controller/BasketController.php [hunk 7]
```diff
@@ -20,7 +20,7 @@
 use Sonata\Component\Basket\BasketFactoryInterface;
 use Sonata\Component\Customer\AddressInterface;
 use Sonata\Component\Delivery\UndeliverableCountryException;
-use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController as Controller;
```
Comment (wbloszyk) [inline src/BasketBundle/Controller/BasketController.php]: Do not change it. It work diffrence. Add it to master with BlockService update, please.

File: src/ProductBundle/Twig/Extension/ProductExtension.php [hunk 8]
```diff
@@ -20,11 +20,12 @@
 use Symfony\Component\Form\Extension\Core\Type\FormType;
 use Symfony\Component\Form\FormFactoryInterface;
 use Symfony\Component\Form\FormView;
+use Twig\Extension\AbstractExtension;
 
 /**
  * @author Sylvain Deloux <sylvain.deloux@ekino.com>
  */
-class ProductExtension extends \Twig_Extension
+class ProductExtension extends AbstractExtension
```
Comment (wbloszyk) [inline src/ProductBundle/Twig/Extension/ProductExtension.php]: Also add it in master.

File: src/ProductBundle/Resources/config/routing/catalog.xml [hunk 9]
```diff
@@ -4,8 +4,14 @@
         <default key="_controller">SonataProductBundle:Catalog:index</default>
     </route>
     <route id="sonata_catalog_category" path="/{category_slug}/{category_id}">
+        <requirement key="category_id">\d+</requirement>
         <default key="_controller">SonataProductBundle:Catalog:index</default>
         <default key="category_id">0</default>
         <default key="category_slug">all</default>
     </route>
+    <route id="sonata_catalog_collection" path="/collection/{collection_slug}/{collection_id}">
```
Comment (wbloszyk) [inline src/ProductBundle/Resources/config/routing/catalog.xml]:
```suggestion
    <route id="sonata_catalog_collection" path="/collection/{collection_slug}/{collection_id}">
        <requirement key="collection_id">\d+</requirement>
```

File: composer.json [hunk 10]
```diff
@@ -18,6 +18,7 @@
         "php": "^7.1",
         "cocur/slugify": "^2.5 || ^3.0",
         "doctrine/doctrine-bundle": "^1.0",
+        "knplabs/knp-components": "^1.3",
```
Comment (wbloszyk) [inline composer.json]: It is required?

File: composer.json [hunk 11]
```diff
@@ -26,10 +27,10 @@
         "sonata-project/classification-bundle": "^3.3",
         "sonata-project/comment-bundle": "^3.1",
         "sonata-project/core-bundle": "^3.15.1",
-        "sonata-project/datagrid-bundle": "^2.3",
+        "sonata-project/datagrid-bundle": "^2.5",
         "sonata-project/doctrine-extensions": "^1.5",
         "sonata-project/doctrine-orm-admin-bundle": "^3.2",
-        "sonata-project/formatter-bundle": "^3.4 || ^4.0",
+        "sonata-project/formatter-bundle": "^4.0",
```
Comment (wbloszyk) [inline composer.json]: It is required?

File: UPGRADE-3.x.md [hunk 12]
```diff
@@ -4,6 +4,8 @@ UPGRADE 3.x
 UPGRADE FROM 3.0 to 3.1
 =======================
 
+## Fix products collection nagivation
```
Comment (core23) [inline UPGRADE-3.x.md]: Either remove this headlines or add some more information to this.

Comment (greg0ire) [reply UPGRADE-3.x.md]: Also, "navigation" is spelt wrong :P

Comment (salamichel) [reply UPGRADE-3.x.md]: Okay guys i'm not english fluent ^^

File: src/BasketBundle/Block/BasketBlockService.php (hunk 6 above)
Comment (salamichel) [reply src/BasketBundle/Block/BasketBlockService.php]: i just leave as i was. see next time

File: UPGRADE-3.x.md (hunk 12 above)
Comment (salamichel) [reply UPGRADE-3.x.md]: @core23 is it better ?

File: src/ProductBundle/Controller/CollectionController.php [hunk 13]
```diff
@@ -21,92 +25,105 @@
...
-     * @param $collectionId
-     * @param $slug
-     *
-     * @throws NotFoundHttpException
+     * List the product related to one collection.
      *
      * @return Response
      */
-    public function viewAction($collectionId, $slug)
+    public function indexAction(Request $request)
```
Comment (core23) [inline src/ProductBundle/Controller/CollectionController.php]: Technically this is a BC break. Can you please create proxy methods, that are calling the new methods?

File: src/ProductBundle/Entity/ProductManager.php [hunk 14]
```diff
@@ -162,6 +162,21 @@ public function getPager(array $criteria, $page, $limit = 10, array $sort = [])
         return $pager;
     }
 
+    public function queryInCollection($collection, $limit = null)
```
Comment (core23) [inline src/ProductBundle/Entity/ProductManager.php]:
```suggestion
    public final function queryInCollection($collection, $limit = null)
```

File: tests/ProductBundle/Controller/CollectionControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+        $productSetManager = $this->createMock(ProductSetManager::class);
+        $productSetManager->expects($this->any())->method('queryincollection')->willReturn(true);
+
+        $container = $this->createMock(ContainerInterface::class);
+        $template = $this->createMock(Environment::class);
+
+        $collectionManagerInterface = $this->createMock(CollectionManagerInterface::class);
+        $collectionManagerInterface->expects($this->any())->method('findOneBy')->willReturn($collection);
+
+        /* $container->expects($this->at(2))
```
Comment (core23) [inline tests/ProductBundle/Controller/CollectionControllerTest.php]: Please don't use commented out code

File: src/ProductBundle/Controller/CollectionController.php [hunk 16]
```diff
@@ -21,92 +25,114 @@
...
+
+    public function __construct(SeoPage $sonataSeoPage = null, CurrencyDetectorInterface $currencyDetector = null, CollectionManagerInterface $collectionManagerInterface = null, ProductSetManager $productSetManager = null)
+    {
+        $this->sonataSeoPage = $sonataSeoPage;
+        $this->currencyDetector = $currencyDetector;
+        $this->collectionManagerInterface = $collectionManagerInterface;
+        $this->productSetManager = $productSetManager;
+    }
+
+    public function viewAction($collectionId, $slug)
```
Comment (core23) [inline src/ProductBundle/Controller/CollectionController.php]: Can you please deprecate all *old* methods

**Comment by salamichel** [comment]
@core23 Hi is there any pending issue?

File: src/ProductBundle/Controller/CollectionController.php [hunk 17]
```diff
@@ -92,21 +152,39 @@ public function listProductsAction(Request $request, $collectionId)
...
     }
 
-    /**
-     * @param null $collection
-     * @param int  $depth
-     * @param int  $deep
-     *
-     * @return Response
-     */
-    public function listSideMenuCollectionsAction($collection = null, $depth = 1, $deep = 0)
```
Comment (core23) [inline src/ProductBundle/Controller/CollectionController.php]: Removing a public method is still a BC break

File: src/ProductBundle/Controller/CollectionController.php [hunk 18]
```diff
@@ -109,4 +173,40 @@ public function listSideMenuCollectionsAction($collection = null, $depth = 1, $d
           'deep' => $deep + 1,
         ]);
     }
+
+    protected function getCurrencyDetector(): CurrencyDetectorInterface
```
Comment (core23) [inline src/ProductBundle/Controller/CollectionController.php]: Either make this final or private

Comment (salamichel) [reply src/ProductBundle/Controller/CollectionController.php]: i don't understand why only this method but it's done

**Comment by salamichel** [comment]
@core23 @greg0ire hi happy new year. Any update?

File: src/ProductBundle/Controller/CollectionController.php [hunk 19]
```diff
@@ -21,46 +25,98 @@
...
+     * @var CollectionManagerInterface
+     */
+    private $collectionManagerInterface;
+
+    /**
+     * @var ProductSetManager
+     */
+    private $productSetManager;
+
+    public function __construct(SeoPage $sonataSeoPage = null, CurrencyDetectorInterface $currencyDetector = null, CollectionManagerInterface $collectionManagerInterface = null, ProductSetManager $productSetManager = null)
```
Comment (core23) [inline src/ProductBundle/Controller/CollectionController.php]: Please break this

File: src/ProductBundle/Controller/CollectionController.php [hunk 20]
```diff
@@ -109,4 +173,40 @@ public function listSideMenuCollectionsAction($collection = null, $depth = 1, $d
           'deep' => $deep + 1,
         ]);
     }
+
+    protected function getCollectionManagerInterface(): CollectionManagerInterface
```
Comment (core23) [inline src/ProductBundle/Controller/CollectionController.php]: please make this private or final (same for the other methods)

File: src/ProductBundle/Controller/CollectionController.php [hunk 21]
```diff
@@ -21,46 +25,98 @@
...
+
+        return $this->indexAction($request);
+    }
+
+    /**
+     * List the product related to one collection.
+     *
+     * @return Response
+     */
+    public function indexAction(Request $request)
```
Comment (core23) [inline src/ProductBundle/Controller/CollectionController.php]:
```suggestion
    public final function indexAction(Request $request)
```

File: src/ProductBundle/Controller/CollectionController.php (hunk 19 above)
Comment (salamichel) [reply src/ProductBundle/Controller/CollectionController.php]: Sorry what do you mean?

Comment (wbloszyk) [reply src/ProductBundle/Controller/CollectionController.php]:
```suggestion
    public function __construct(
        SeoPage $sonataSeoPage = null, 
        CurrencyDetectorInterface $currencyDetector = null, 
        CollectionManagerInterface $collectionManagerInterface = null, 
        ProductSetManager $productSetManager = null
    )
```
It is more readeable

**Comment by salamichel** [comment]
Hi Still ko? is there a wiki page for best pratice ?

File: src/ProductBundle/Controller/CollectionController.php [hunk 22]
```diff
@@ -21,51 +25,107 @@
...
+
+        return $this->indexAction($request);
+    }
+
+    /**
+     * List the product related to one collection.
+     *
+     * @return Response
+     */
+    final public function indexAction(Request $request)
```
Comment (greg0ire) [inline src/ProductBundle/Controller/CollectionController.php]:
```suggestion
    final public function indexAction(Request $request): Response
```

File: src/ProductBundle/Controller/CollectionController.php [hunk 23]
```diff
@@ -21,51 +25,107 @@
...
+        $request->query->set('collection_id', $collectionId);
+        $request->query->set('collection_slug', $slug);
+
+        return $this->indexAction($request);
+    }
+
+    /**
+     * List the product related to one collection.
+     *
+     * @return Response
```
Comment (greg0ire) [inline src/ProductBundle/Controller/CollectionController.php]:
```suggestion
```

File: src/ProductBundle/Controller/CollectionController.php [hunk 24]
```diff
@@ -21,51 +25,107 @@
...
+        $request = new Request();
+        $request->query->set('collection_id', $collectionId);
+        $request->query->set('collection_slug', $slug);
+
+        return $this->indexAction($request);
+    }
+
+    /**
+     * List the product related to one collection.
+     *
```
Comment (greg0ire) [inline src/ProductBundle/Controller/CollectionController.php]:
```suggestion
```

File: src/ProductBundle/Entity/ProductManager.php [hunk 25]
```diff
@@ -162,6 +162,21 @@ public function getPager(array $criteria, $page, $limit = 10, array $sort = [])
         return $pager;
     }
 
+    final public function queryInCollection($collection, $limit = null)
```
Comment (greg0ire) [inline src/ProductBundle/Entity/ProductManager.php]: Please add a return type declaration

File: tests/CustomerBundle/Block/RecentCustomersBlockServiceTest.php [hunk 26]
```diff
@@ -25,9 +25,11 @@ class RecentCustomersBlockServiceTest extends TestCase
 {
     public function testGetName(): void
     {
+        $environment = $this->createMock(\Twig\Environment::class);
```
Comment (greg0ire) [inline tests/CustomerBundle/Block/RecentCustomersBlockServiceTest.php]: Please use a use statement for this

**Comment by greg0ire** [comment]
Thanks @salamichel !

**Pull request merged**